File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int size = Integer.parseInt(br.readLine());
        Stack<Integer> stack = new Stack<>();
        int[] originalArray = new int[size];
        int[] sortedArray = new int[size];
        int pos = 0;

        for (int i = 0; i < size; i++) {
            originalArray[i] = Integer.parseInt(br.readLine());
        }
        sortedArray = originalArray.clone();
        Arrays.sort(sortedArray);

        for (int i : sortedArray) {
            stack.push(i);
            sb.append("+\n");
            if (!stack.isEmpty()) {
                while(stack.peek() == originalArray[pos]) {
                    stack.pop();
                    pos++;
                    sb.append("-\n");
                    if (stack.isEmpty()) break;
                }
            }
        }
        if (stack.isEmpty()) {
            System.out.print(sb);
        } else {
            System.out.print("NO");
        }
        br.close();
    }
}

File: 백준/Silver/1874. 스택 수열/스택 수열.kt
import java.io.*
import java.lang.StringBuilder
import java.util.*

fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val stringBuilder = StringBuilder()
    val stack = Stack<Int>()
    val size = br.readLine().toInt()
    var num = mutableListOf<Int>()

    repeat(size) {
        num.add(br.readLine().toInt())
    }

    for (i in 1..size) {
        stack.push(i)
        stringBuilder.append("+\n")
        while(true) {
            if (stack.size > 0) {
                if (stack.peek() == num[0]) {
                    stack.pop()
                    if (num.isNotEmpty()) {
                        num.removeAt(0)
                    }
                    stringBuilder.append("-\n")
                } else if (stack.peek() > num[0]) {
                    println("NO")
                    return
                } else {
                    break
                }
            } else break
        }
    }
    print(stringBuilder)
}